eudemo/eudemo/maintenance/upper_port.py

```diff
@@ -59,127 +54,74 @@ class UpperPortOPParameters(ParameterFrame):
-        n_constraints = 3
+    def port_size(self, x: np.ndarray) -> float:
+        """Return the port size given parameterisation ``x``."""
+        ri, ro, _, gamma = x
+        return ro - ri + gamma
 
-        constraints = [
-            OptimisationConstraint(
-                self.constrain_blanket_cut,
-                f_constraint_args={
-                    "bb": breeding_blanket_xz,
-                    "c_rm": c_rm,
-                    "r_ib_min": r_ib_min,
-                    "r_ob_max": r_ob_max,
-                },
-                tolerance=constraint_tol * np.ones(n_constraints),
-            )
-        ]
-        super().__init__(np.array([]), optimiser, objective, constraints)
+    def df_port_size(self, _: np.ndarray) -> np.ndarray:
+        """
+        Return the gradient of the port size.
 
-        lower_bounds = [r_ib_min - c_rm, R_0, r_ib_min + tk_bb_ib, 0]
-        upper_bounds = [R_0, r_ob_max + c_rm, r_ob_max - tk_bb_ob, bb_min_angle]
-        self.set_up_optimiser(n_variables, bounds=[lower_bounds, upper_bounds])
-        self.params = params
+        Parameters
+        ----------
+        _:
+            The parameterisation of the port size. This is unused as the
+            gradient is constant.
 
-    @staticmethod
-    def minimise_port_size(vector, grad):
+        Returns
+        -------
+        The gradient of the port size parameterisation, with shape (3,).
         """
-        Minimise the size of the port.
-        """
-        ri, ro, ci, gamma = vector
-        # Dual objective: minimise port (ro - ri) and minimise cut angle
-        value = ro - ri + gamma
-        if grad.size > 0:
-            grad[0] = -1
-            grad[1] = 1
-            grad[2] = 0
-            grad[3] = 1
-        return value
+        return self.gradient
 
-    @staticmethod
-    def constrain_blanket_cut(constraint, vector, grad, bb, c_rm, r_ib_min, r_ob_max):
-        """
-        Calculate the constraint and grad matrices.
+    def constrain_blanket_cut(self, x: np.ndarray) -> np.ndarray:
         """
-        constraint[:] = UpperPortOP.calculate_constraints(
-            vector, bb, c_rm, r_ib_min, r_ob_max
-        )
-
-        if grad.size > 0:
-            grad[:] = approx_derivative(
-                UpperPortOP.calculate_constraints,
-                vector,
-                f0=constraint,
-                args=(bb, c_rm, r_ib_min, r_ob_max),
-            )
+        Constrain the upper port size.
 
-        return constraint
+        This enforces 3 constraints:
 
-    @staticmethod
-    def calculate_constraints(vector, bb, c_rm, r_ib_min, r_ob_max):
+        c1. The outboard blanket must fit through the port.
+        c2. The inboard blanket must squeeze past the other blanket
+            segment. Note that this also enforces that the inboard
+            blanket fits through the port.
+        c3. There should be enough vertically accessible space on the
+            inboard blanket.
         """
```

Comment:
On the upper surface of the inboard blanket (i.e. the first thing you see when standing at the top of the upper port, looking down, you should have enough room on the inboard blanket to connect pipes and a RM attachment point. I can't remember exactly how I tried to constrain this in the end.

Here was my sketch of the problem many moons ago (not sure it is relevant or helpful, but at least it doesnt live in my notebook any more):
![image](https://github.com/Fusion-Power-Plant-Framework/bluemira/assets/26097289/d13f9a91-9a43-4674-8ce4-ebd84a68ead2)